runtime/src/utils/buffer/write.rs

```diff
@@ -2,104 +2,116 @@ use crate::{Blob, Error, RwLock};
-    /// The buffer is maintained at the "tip" to support efficient size calculation and appends.
-    position: u64,
+    /// This represents the logical position in the blob where `data[0]` would be written. The
+    /// buffer is maintained at the "tip" to support efficient size calculation and appends.
+    offset: u64,
     /// The maximum size of the buffer.
     capacity: usize,
 }
 
-impl<B: Blob> Inner<B> {
+impl Buffer {
     /// Returns the current logical size of the blob including any buffered data.
     fn size(&self) -> u64 {
-        self.position + self.buffer.len() as u64
+        self.offset + self.data.len() as u64
     }
 
-    /// Appends bytes to the internal buffer, maintaining the "buffer at tip" invariant.
+    /// Truncates the buffer to the provided `len` if it falls within the current range.
+    /// Returns `false` and resets the buffer if the truncation point precedes the current range.
     ///
-    /// If the buffer capacity would be exceeded, it is flushed first. If the data
-    /// is larger than the buffer capacity, it is written directly to the blob.
+    /// # Panics
     ///
-    /// Returns an error if the write to the underlying [Blob] fails (may be due to a `flush` of data not
-    /// related to the data being written).
-    async fn write<S: Into<StableBuf>>(&mut self, buf: S) -> Result<(), Error> {
-        // If the buffer capacity will be exceeded, flush the buffer first
-        let buf = buf.into();
-        let buf_len = buf.len();
-
-        // Flush buffer if adding this data would exceed capacity
-        if self.buffer.len() + buf_len > self.capacity {
-            self.flush().await?;
-        }
-
-        // Write directly to blob if data is larger than buffer capacity
-        if buf_len > self.capacity {
-            self.blob.write_at(buf, self.position).await?;
-            self.position += buf_len as u64;
-            return Ok(());
+    /// Panics if the truncation point is greater than the current size of the blob.
+    fn truncate(&mut self, len: u64) -> bool {
+        assert!(len <= self.size());
+        if len <= self.offset {
+            self.data.clear();
+            self.offset = len;
+            return false;
         }
+        self.data.truncate((len - self.offset) as usize);
+        true
+    }
 
-        // Append to buffer (buffer is now guaranteed to have space)
-        self.buffer.extend_from_slice(buf.as_ref());
-        Ok(())
+    /// Returns true if the buffer is empty.
+    fn is_empty(&self) -> bool {
+        self.data.is_empty()
     }
 
-    /// Flushes buffered data to the underlying [Blob] and advances the position.
+    /// Takes & returns the buffered data for flushing, setting up the buffer for a new write.
+    ///
+    /// # Panics
     ///
-    /// After flushing, the buffer is reset and positioned at the new tip.
-    /// Does nothing if the buffer is empty.
+    /// Panics if the buffer is empty.
```

Comment:
I wonder if this assertion is too strict 🤔

Reply:
It is obviously fine in how we use it, but not sure if we should assert for any use (I suppose the invariant is you should always make sure non-empty before taking).

Reply:
See my note below about handling this slightly differently.